common/spec/specfiles.go

```diff
@@ -204,6 +205,9 @@ func ValidateSpec(files []File, isTargetMandatory, isSearchBasedSpec, isUpload b
 					return fileSpecValidationError("bundle", "offset")
 				}
 			}
```

Comment:
The following suggested change nay have no impact on the validation functionality, but it'll probably make the code look better.
The only if statement that should remain inside the following statement -
```
if isSearchBasedSpec {
```
is this - 
```
	if !isAql && !isPattern && !isBuild && !isBundle {
		return errors.New("Spec must include either aql, pattern, build or bundle.")
	}
```
The rest should be moved out of it, and put right after it. 
I don't see how ```isSearchBasedSpec``` is relevant for anything else, except for the above check.
Actually, I suggest the modifying the function to be - 
```
func ValidateSpec(files []File, isTargetMandatory, isSearchBasedSpec bool) error {
	if len(files) == 0 {
		return errors.New("Spec must include at least one file group")
	}

	for _, file := range files {
		isAql := len(file.Aql.ItemsFind) > 0
		isPattern := len(file.Pattern) > 0
		isExclusions := len(file.Exclusions) > 0 && len(file.Exclusions[0]) > 0
		isTarget := len(file.Target) > 0
		isSortOrder := len(file.SortOrder) > 0
		isSortBy := len(file.SortBy) > 0
		isBuild := len(file.Build) > 0
		isExcludeArtifacts, _ := file.IsExcludeArtifacts(false)
		isIncludeDeps, _ := file.IsIncludeDeps(false)
		isBundle := len(file.Bundle) > 0
		isGPGKey := len(file.PublicGpgKey) > 0
		isOffset := file.Offset > 0
		isLimit := file.Limit > 0
		isValidSortOrder := file.SortOrder == "asc" || file.SortOrder == "desc"
		isExcludeProps := len(file.ExcludeProps) > 0
		isArchive := len(file.Archive) > 0
		isValidArchive := file.Archive == "zip"
		isSymlinks, _ := file.IsSymlinks(false)
		isRegexp := file.Regexp == "true"
		isAnt := file.Ant == "true"
		isExplode, _ := file.IsExplode(false)
		isTransitive, _ := file.IsTransitive(false)

		if isTargetMandatory && !isTarget {
			return errors.New("Spec must include target.")
		}
		if !isSearchBasedSpec && !isPattern {
			return errors.New("Spec must include a pattern.")
		}
		if isBuild && isBundle {
			return fileSpecValidationError("build", "bundle")
		}
		if isSearchBasedSpec && !isAql && !isPattern && !isBuild && !isBundle {
			return errors.New("Spec must include either aql, pattern, build or bundle.")
		}
		if isOffset {
			if isBuild {
				return fileSpecValidationError("build", "offset")
			}
			if isBundle {
				return fileSpecValidationError("bundle", "offset")
			}
		}
		if isTransitive && isOffset {
			return fileSpecValidationError("transitive", "offset")
		}
		if isLimit {
			if isBuild {
				return fileSpecValidationError("build", "limit")
			}
			if isBundle {
				return fileSpecValidationError("bundle", "limit")
			}
		}
		if isAql && isPattern {
			return fileSpecValidationError("aql", "pattern")
		}
		if isAql && isExclusions {
			return fileSpecValidationError("aql", "exclusions")
		}
		if isAql && isExcludeProps {
			return fileSpecValidationError("aql", "excludeProps")
		}
		if !isSortBy && isSortOrder {
			return errors.New("Spec cannot include 'sort-order' if 'sort-by' is not included")
		}
		if isSortOrder && !isValidSortOrder {
			return errors.New("The value of 'sort-order' can only be 'asc' or 'desc'.")
		}
		if isTransitive && isSortBy {
			return fileSpecValidationError("transitive", "sort-by")
		}
		if !isBuild && (isExcludeArtifacts || isIncludeDeps) {
			return errors.New("Spec cannot include 'exclude-artifacts' or 'include-deps' if 'build' is not included.")
		}
		if isRegexp && isAnt {
			return errors.New("Can not use the option of regexp and ant together.")
		}
		if isArchive && isSymlinks && isExplode {
			return errors.New("Symlinks cannot be stored in an archive that will be exploded in artifactory.\\nWhen uploading a symlink to Artifactory, the symlink is represented in Artifactory as 0 size filewith properties describing the symlink.\\nThis symlink representation is not yet supported by Artifactory when exploding symlinks from a zip.")
		}
		if isArchive && !isValidArchive {
			return errors.New("The value of 'archive' (if provided) must be 'zip'.")
		}
		if isGPGKey && !isBundle {
			return errors.New("Spec cannot include 'gpg-key' if 'bundle' is not included.")
		}
	}
	return nil
}
```